lib/Parse/Parser.cpp

```diff
@@ -982,13 +980,40 @@ bool Parser::parseMatchingToken(tok K, SourceLoc &TokLoc, Diag<> ErrorDiag,
   if (parseToken(K, TokLoc, ErrorDiag)) {
     diagnose(OtherLoc, OtherNote);
 
-    TokLoc = PreviousLoc;
+    TokLoc = getLocForMissingMatchingToken();
     return true;
   }
 
   return false;
 }
 
+SourceLoc Parser::getLocForMissingMatchingToken() const {
+  // The right brace, parenthesis, etc. must include the whole of the previous
+  // token in order so that an unexpanded lazy \c IterableTypeScope includes its
+  // contents.
+  return getErrorOrMissingLoc();
+}
+
+SourceLoc Parser::getErrorOrMissingLoc() const {
+  // LazyASTScopes require that the ends of types, etc. which are at the end of
+  // \c IterableTypeContext \c Decls, such as \c StructDecl not be beyond the
+  // end of the enclosing \c Decl.
+
+  // The missing right brace, parenthesis, etc. must include the whole of the
+  // previous token in order so that an unexpanded lazy \c IterableTypeScope
+  // includes its contents.
+
+  auto const PreviousTok = Lexer::getTokenAtLocation(Context.SourceMgr,
+                                                     PreviousLoc);
+  // If it's a string literal, it might be an InterpolatedStringLiteral.
+  // In that case the missing close paren, etc. needs to go at the end
+  // because there might be things to be looked up for ASTScope in the
+  // middle.
+  return PreviousTok.getKind() != tok::string_literal
+    ? PreviousLoc
+    : PreviousLoc.getAdvancedLoc(PreviousTok.getLength() - 1);
```

Comment:
Same issue here, this is returning a character-based location. I think this should always return `PreviousLoc`, which AFAICT is the location of the last parsed token.

Reply:
I can do this, and try the change suggested to ASTScopes (using character locations). And I'm happy to give it a whirl. But I'm still bothered because the notion of "last parsed token" seems ill-defined to me, as if we're building on sand.

Consider `"\(foo)"`. You could say that the last parsed token is the whole interpolated string literal, except that the last parsed token is also the closing parenthesis, or maybe the `foo`.
Before I added the save-and-restore line for `PreviousLoc` that value did point to the `foo` IIRC. Part of the confusion is in the word "last", I suppose. Is that the most-recently-parsed? (in which case the ambiguity is whether it's the most-recently-started-to-parse or the most-recently-finished-to-parse). Or is it the last token that has been parsed in the file? (I.e., the token furthest from the start of the file that has been parsed?)

If you help me understand this, I'll try to add some comments, too.

Reply:
The thing the complicates the interpolated string literal is that it is a "tokens within a token" situation, but a way to reason about it is that there are these set of tokens:
* the string literal token `"\(foo)"`
* The interpolated tokens `'(', 'foo', ')'`

Then the question is what kind of source range you want. If you want the range of the whole string literal then you should use the literal token range. If you want the range of the interpolated segment then you should use the interpolated tokens.

But you cannot have the `EndLoc` of the SourceRange be the closing quote, because the closing quote is not a token, and the SourceRange you'd form with it would not be token-based.

Reply:
Thanks, that helps. Given that, if I want to keep ASTScope token-like SourceRange based, the key would be to look inside the string literal, at the tokens of the expression within, and take the location of the last of those tokens, the ')'. Or as you say, change ASTScopes to be character-location based.

The other question is where should the EndLoc of a StructDecl (etc.) be when the close brace is missing? Given the rule that SourceLocs refer to tokens, and that sub-nodes are enclosed (perhaps improperly) by super-nodes, one could put that EndLoc either at the SourceLoc of the string literal (i.e. the open '"'), or at the SourceLoc of the ')'. Do you have a sense of which of these would be more fitting? What's tough here is that we're talking about illegal code.

Reply:
Going to try using the SourceLoc of the string literal, the open quote.